### PR TITLE
Added customizeDiff for a couple of subfields of params field in google_bigquery_data_transfer_config

### PR DESCRIPTION
**Pull request by AlfatahB**
Added customizeDiff for `data_path_template` and `destination_table_name_template` subfields of params field in google_bigquery_data_transfer_config
fixes https://github.com/hashicorp/terraform-provider-google/issues/6992

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
bigquerydatatransfer: recreate `google_bigquery_data_transfer_config` for Cloud Storage transfers when immutable params `data_path_template` and `destination_table_name_template` are changed
```

### CODE REVIEW

**Comment by AlfatahB** [comment]
I've tried to modify the value of the `destination_table_name_template` field. I was getting the error stating `Immutable parameter destination_table_name_template with value string_value: \t \"the-table\"\n cannot be changed to string_value: \t \"the-table-new\"\n` as shown in the attached screenshot. This was the same error I was getting while changing the value of `data_path_template`. Hence, I used customizeDiff to call `diff.ForceNew("params")`  if either of them have changed in value.
Screenshot from 2022-10-31 15-23-25

**Comment by melinath** [comment]
assigning to @roaks3 to do a preliminary review

**Comment by roaks3** [comment]
This PR looks good to me, but two comments:
- I noticed that this looks to be the same implementation that was previously reverted. I see the request/response that seems to confirm that `destination_table_name_template` is immutable, but I'm wondering why we believed it was not when we reverted. Is there any way we can confirm that this won't re-trigger a case where we need to revert?
- Since we are implementing a particular ForceNew behavior here, I think it would be best to include a simple unit test like this: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/tests/resource_spanner_database_test.go.erb#L375-L450. That way we can confirm these two fields trigger a ForceNew, and the other fields within `params` do not. However, since the previous behavior here was a failure if you tried to update, I don't think it's strictly necessary.

**Comment by melinath** [comment]
> I'm wondering why we believed it was not when we reverted.

@c2thorn may have context related to this.

**Comment by AlfatahB** [comment]
> * Since we are implementing a particular ForceNew behavior here, I think it would be best to include a simple unit test like this: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/tests/resource_spanner_database_test.go.erb#L375-L450. That way we can confirm these two fields trigger a ForceNew, and the other fields within `params` do not. However, since the previous behavior here was a failure if you tried to update, I don't think it's strictly necessary.

I have added the unit tests for the particular ForceNew behaviour.

**Comment by c2thorn** [comment]
> > I'm wondering why we believed it was not when we reverted.
> 
> @c2thorn may have context related to this.

I'm still able to update this field successfully via the provider. I believe this field may have conditional mutability. I've attached my logs at b/251556517

**Comment by AlfatahB** [comment]
Thanks @c2thorn for the inputs. I've found out that if `data_source_id` is set to "google_cloud_storage" than `data_path_template` and `destination_table_name_template` are immutable fields. So, I've added a condition in customizeDiff to check if the value of `data_source_id` is set to "google_cloud_storage" then we have to call `diff.ForceNew("params")`.

**Review by roaks3 — APPROVED** [review]
For reference, I did find this line in the [docs](https://cloud.google.com/bigquery/docs/cloud-storage-transfer) that appears to confirm the limitation with Cloud Storage (at least for `destination_table`):

> Unlike [individual loads of data from Cloud Storage to BigQuery](https://cloud.google.com/bigquery/docs/loading-data-cloud-storage), for ongoing transfers you need to [create the destination table and its schema](https://cloud.google.com/bigquery/docs/schemas) in advance of setting up the transfer. BigQuery cannot create the table as part of the recurring data transfer process.

I think this is definitely service-specific logic that we prefer not to re-implement in the provider, but I'm unsure of a better way to solve the problem. For that reason, I believe this custom ForceNew logic is our best option, and moving forward we would need to keep an eye out for other data sources that need similar adjustments.

Adding Stephen for follow-up review.

File: mmv1/templates/terraform/constants/bigquery_data_transfer.go.erb [hunk 1]
```diff
@@ -9,4 +9,39 @@ func sensitiveParamCustomizeDiff(_ context.Context, diff *schema.ResourceDiff, v
 		}
 	}
 	return nil
-}
\ No newline at end of file
+}
+
+// This customizeDiff is to use ForceNew for params fields data_path_template and 
+// destination_table_name_template only if the value of "data_source_id" is "google_cloud_storage".
+// data_path_template and destination_table_name_template are immutable for "google_cloud_storage".
```
Comment (melinath) [inline mmv1/templates/terraform/constants/bigquery_data_transfer.go.erb:16]: This line is redundant

```suggestion
```

File: mmv1/third_party/terraform/tests/resource_bigquery_data_transfer_config_test.go [hunk 2]
```diff
@@ -10,6 +10,106 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 )
 
+func TestBigqueryDataTransferConfig_resourceBigqueryDTCParamsCustomDiffFuncForceNew(t *testing.T) {
```
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_bigquery_data_transfer_config_test.go:13]: please add cases to test that the resource won't be recreated if any of these fields change but data_source_id is _not_ google_cloud_storage. (That is, make copies of the test cases but change data_source_id and have them all expect forcenew: false)

**Review by melinath — CHANGES_REQUESTED** [review]

File: mmv1/templates/terraform/constants/bigquery_data_transfer.go.erb (hunk 1 above)
Comment (AlfatahB) [reply mmv1/templates/terraform/constants/bigquery_data_transfer.go.erb]: Okay, I've removed this line.

File: mmv1/third_party/terraform/tests/resource_bigquery_data_transfer_config_test.go (hunk 2 above)
Comment (AlfatahB) [reply mmv1/third_party/terraform/tests/resource_bigquery_data_transfer_config_test.go:13]: `scheduled_query` data_source_id doesn't expect the `data_path_template` field in the `params` field. So, I've only added a test case for the `destination_table_name_template` field where data_source_id is `scheduled_query`.

**Review by melinath — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/tests/resource_bigquery_data_transfer_config_test.go [hunk 3]
```diff
@@ -10,6 +10,125 @@ import (
...
+					"data_path_template":              "gs://bq-bucket-temp/*.json",
+					"destination_table_name_template": "table-old",
+					"file_format":                     "JSON",
+					"max_bad_records":                 1000,
+					"write_disposition":               "APPEND",
+				},
+			},
+			forcenew: false,
+		},
+		"changing_destination_table_name_template_for_different_data_source_id": {
```
[the remark this answers is not in the document]
Comment (AlfatahB) [reply mmv1/third_party/terraform/tests/resource_bigquery_data_transfer_config_test.go:90]: Okay, I've added a test case for `scheduled_query` + `data_path_template`

**Review by melinath — APPROVED** [review]

**Pull request merged**